File: scala/P04.scala
// P04 Find the number of elements of a list.

// Predefined function
def f0[T](list: List[T]): Int = list.length

// Recursion
def f1[T](list: List[T]): Int = list match {
  case Nil => 0
  case _ :: t => 1 + f1(t)
}

// Tail recursion
@annotation.tailrec
def f2[T](list: List[T], n: Int = 0): Int = list match {
  case Nil => n
  case _ :: t => f2(t, n + 1)
}

// Folding
def f3[T](list: List[T]): Int = list.foldLeft(0)((a, _) => a + 1)

// Using build-in 'sum'
def f4[T](list: List[T]): Int = list.map(_ => 1).sum

// With 'view' to avoid the creation of a temporary list
def f5[T](list: List[T]): Int = list.view.map(_ => 1).sum

File: haskell/p04.hs
-- P04 Find the number of elements of a list.

module P04 where

-- Predefined function
f0 :: [a] -> Int
f0 = length

-- Recursion
f1 :: [a] -> Int
f1 [] = 0
f1 (_:xs) = 1 + f1 xs

-- Tail recursion
f2 :: [a] -> Int
f2 = f2' 0
    where f2' :: Int -> [a] -> Int
          f2' n [] = n
          f2' n (_:xs) = f2' (n + 1) xs

-- Folding
f3 :: [a] -> Int
f3 xs = foldl (\a _ -> a + 1) 0 xs

-- Folding with function composition
f4 :: [a] -> Int
f4 = foldr (const (+1)) 0

-- Using build-in 'sum'
f5 :: [a] -> Int
f5 = sum . map (const 1)